scala-package/core/src/test/scala/org/apache/mxnet/ResourceScopeSuite.scala

```diff
@@ -0,0 +1,151 @@
+      */
+    override def nativeDeAllocator: CPtrAddress => Int = TestNativeResource.deAllocator
+
+    /** Call NativeResource.register to get the reference
+      */
+    override val ref: NativeResourceRef = super.register()
+    /**
+      * Off-Heap Bytes Allocated for this object
+      */
+    override val bytesAllocated: Long = 0
+  }
+  object TestNativeResource {
+    def deAllocator(handle: CPtrAddress): Int = 0
+  }
+
+  object TestPhantomRef  {
+    def getRefQueue: ReferenceQueue[NativeResource] = { NativeResourceRef.refQ}
+    def getRefMap: ConcurrentHashMap[NativeResourceRef, CPtrAddress]
+    = {NativeResourceRef.refMap}
+    def getCleaner: Thread = { NativeResourceRef.cleaner }
+
+  }
+
+  class TestPhantomRef(resource: NativeResource,
+                       resourceDeAllocator: CPtrAddress => Int)
+    extends NativeResourceRef(resource, resourceDeAllocator) {
+  }
+
+  test(testName = "test NDArray Auto Release") {
+    var a: NDArray = null
+    var aRef: NativeResourceRef = null
+    var b: NDArray = null
+
+    ResourceScope.using() {
+      b = ResourceScope.using() {
+          a = NDArray.ones(Shape(3, 4))
+          aRef = a.ref
+          val x = NDArray.ones(Shape(3, 4))
+        x
+      }
+      val bRef: NativeResourceRef = b.ref
+      assert(a.isDisposed == true,
+        "objects created within scope should have isDisposed set to true")
+      assert(b.isDisposed == false,
+        "returned NativeResource should not be released")
+      assert(TestPhantomRef.getRefMap.containsKey(aRef) == false,
+        "reference of resource in Scope should be removed refMap")
+      assert(TestPhantomRef.getRefMap.containsKey(bRef) == true,
+        "reference of resource outside scope should be not removed refMap")
+    }
+    assert(b.isDisposed, "resource returned from inner scope should be released in outer scope")
+  }
+
+  test("test return object release from outer scope") {
+    var a: TestNativeResource = null
+    ResourceScope.using() {
+      a = ResourceScope.using() {
+        new TestNativeResource()
+      }
+      assert(a.isDisposed == false, "returned object should not be disposed within Using")
+    }
+    assert(a.isDisposed == true, "returned object should be disposed in the outer scope")
+  }
+
+  test(testName = "test NativeResources in returned Lists are not disposed") {
+    var ndListRet: IndexedSeq[TestNativeResource] = null
+    ResourceScope.using() {
+      ndListRet = ResourceScope.using() {
+        val ndList: IndexedSeq[TestNativeResource] =
+          IndexedSeq(new TestNativeResource(), new TestNativeResource())
+        ndList
+      }
+      ndListRet.foreach(nd => assert(nd.isDisposed == false,
+        "NativeResources within a returned collection should not be disposed"))
+    }
+    ndListRet.foreach(nd => assert(nd.isDisposed == true,
+    "NativeResources returned from inner scope should be disposed in outer scope"))
+  }
+
+  test("test native resource inside a map") {
```

Comment:
👍🏿

Reply:
thanks for helping with the syntax, need to do the same for Java Iterables.